File: implementations/data structures/queues/queue.swift
//
//  main.swift
//  queues
//
//  Created by Lareen Melo on 2/21/20.
//  Copyright © 2020 Lareen Melo. All rights reserved.
//

import Foundation

struct Queue<T> {
    var q: [T] = []
    
    mutating func enqueue(_ value: T) {
        q.append(value)
    }
    
    mutating func dequeue() -> T? {
        guard q.count != 0 else { return nil }
        guard let element = q.first else { return nil }
        q.removeFirst()
        
        return element
    }
    
    func front() -> T? {
        guard let front = q.first else { return nil }
        
        return front
    }
    
    func printQueue() {
        for item in q {
            print(item)
        }
    }
}


var myQueue = Queue<Int>()
myQueue.enqueue(5)
myQueue.enqueue(4)
myQueue.enqueue(3)
myQueue.enqueue(2)
myQueue.enqueue(1)

print("Deleting \(myQueue.dequeue()!)")
print("Deleting \(myQueue.dequeue()!)")
print("Deleting \(myQueue.dequeue()!)")


var result: Int {
    if let number = myQueue.front() {
        return number
    }
    return -1
    
}

print("Front of list: \(result)")

myQueue.printQueue()

File: implementations/data structures/queues/queue.cpp
//
//  main.cpp
//  stacks
//
//  Created by Lareen Melo on 2/20/20.
//  Copyright © 2020 Lareen Melo. All rights reserved.
//

#include <iostream>

const int CAPACITY = 5;
template <class T>
class Queue {
private:
    int rear;
    int _front;
    T q[CAPACITY];
    
public:
    Queue() {
        rear = -1;
        _front = -1;
    }
    
    void dequeue() {
        if (rear == -1 && _front == -1) {
            std::cout << "underflow" << std::endl;
            return;
        }
        
        std::cout << "Deleting ";
        front();

        if (rear == _front) {
            _front = -1;
            rear = -1;
        } else {
            
            _front++;
        }
    }
    
    void enqueue(T value) {
        if (rear == -1 && _front == -1) {
            rear++;
            _front++;
            
        } else {
            if (rear < CAPACITY-1) {
                rear++;
                
            } else {
                std::cout << "overflow" << std::endl;
                return;
            }
        }
        
        q[rear] = value;
    }
    
    void front() {
        std::cout << q[_front] << std::endl;
    }
    
    void print() {
        if (_front == -1 && rear == -1) {
            std::cout << 0 << std::endl;
            
        } else {
            for(int i = _front; i <= rear; i++) {
                if (i == rear) {
                    std::cout << q[i] << "." << std::endl;
                } else {
                    std::cout << q[i] << ", ";
                }
            }
        }
    }
};


int main() {
    Queue<int> queue;
    
    queue.enqueue(5);
    queue.enqueue(4);
    queue.enqueue(3);
    queue.enqueue(2);
    queue.enqueue(1);
    
    queue.dequeue();
//    queue.dequeue();
//    queue.dequeue();
//    queue.dequeue();
//    queue.dequeue();

    queue.front();
    queue.print();
    return 0;
}

